src/Orleans/Async/BatchWorker.cs

```diff
@@ -0,0 +1,175 @@
+    /// The worker never executes more than one instance of the work cycle at a time, 
+    /// and consumes no resources when idle. It uses TaskScheduler.Current 
+    /// to schedule the work cycles.
+    /// </summary>
+    public abstract class BatchWorker
+    {
+        // Subclass overrides this to define what constitutes a work cycle
+        protected abstract Task Work();
+
+        /// <summary>
+        /// Notify the worker that there is more work.
+        /// </summary>
+        public void Notify()
+        {
+            lock (this)
+            {
+                if (currentworkcycle != null)
+                {
+                    // lets the current work cycle know that there is more work
+                    morework = true;
+                }
+                else
+                {
+                    // start a work cycle
+                    Start();
+                }
+            }
+        }
+
+        // task for the current work cycle, or null if idle
+        private Task currentworkcycle;
+ 
+        // flag is set to indicate that more work has arrived during execution of the task
+        private volatile bool morework;
+
+        // is non-null if some task is waiting for the next work cycle to finish
+        private TaskCompletionSource<Task> nextworkcyclepromise;
+
+        private void Start()
+        {
+            // start the task that is doing the work
+            currentworkcycle = Work();
+
+            // chain a continuation that checks for more work, on the same scheduler
+            currentworkcycle.ContinueWith(t => this.CheckForMoreWork(), TaskScheduler.Current);
+        }
+
+        // executes at the end of each work cycle
+        // on the same task scheduler
+        private void CheckForMoreWork()
+        {
+            Action signal_thunk = null;
+
+            lock (this)
+            {
+                if (morework)
+                {
+                    morework = false;
+
+                    // see if someone created a promise for waiting for the next work cycle
+                    // if so, take it and remove it
+                    var x = this.nextworkcyclepromise;
+                    this.nextworkcyclepromise = null;
+
+                    // start the next work cycle
+                    Start();
+
+                    // if someone is waiting, signal them
+                    if (x != null)
+                        signal_thunk = () => { x.SetResult(currentworkcycle); };
+                }
+                else
+                {
+                    currentworkcycle = null;
+                }
+            }
+
+            // to be safe, must do the signalling out here so it is not under the lock
+            if (signal_thunk != null)
+                signal_thunk();
+        }
+
+        /// <summary>
+        /// Check if this worker is busy.
+        /// </summary>
+        /// <returns></returns>
+        public bool IsIdle()
+        {
+            lock (this)
```

Comment:
Hmm, I am slightly skeptical here. You have any official reference documentation on the C# memory model?
Generally, one should not allow data races on anything that is not volatile. But let's not have this issue get in the way of this PR getting merged ASAP.